### ISSUE TITLE
mu4e: Opening headers view is slow with show-threads and include-related

### ISSUE THREAD

**Issue by tmalsburg** [djcb/mu#163]
My inbox (synced from [Gmail].All Mail) contains ~ 38k messages but finding messages is remarkable fast (I did `time mu ... > /dev/null`):
- `find ""`: 1.9 s
- `find -r ""`: 3.6 s
- `find -t ""`: 5.0 s
- `find -r - t ""`: 6.7 s

However, in mu4e I get much longer times when opening headers view.  When I set `mu4e-headers-results-limit` to 500:
- without threads, without related messages: instantly
- with threads, without related messages: 3 s
- with threads, with related messages: _19 s_

This time is almost completely spent in mu server and not in Emacs.  Interestingly, I get the same times when I set `mu4e-headers-results-limit` to 100.  When I do the same experiment with `mu4e-headers-results-limit` set to 1, I get:
- without threads, without related messages: instantly
- with threads, without related messages: 3 s
- with threads, with related messages: 5 s

Apparently, it takes staggering 2 seconds to retrieve related messages (there were none) for the single message in the last experiment when the results-limit was 1.  Shouldn't these operations be much faster given how fast mu is when called on the command line?  Threads and and the ability to see related messages are very useful.  However, a 19 seconds waiting time just for opening an inbox is hard to swallow. without

**Comment by vspinu** [comment]
I am getting 12 s with threads and 22 with related. 

I cannot define a query in `mu4e-maildir-shortcuts` to speed up things, can I? What's the recommended way to deal with this?

**Comment by djcb** [comment]
Hmmm, that's very slow... can you try e.g.
```mu find "hello" --threads --include-related -n 500```
(the ```-n``` does what ```mu4e-headers-results-limit``` does.

**Comment by vspinu** [comment]
> mu find "hello" --threads --include-related -n 500

That's fasts:

```
real    0m0.608s
user    0m0.449s
sys     0m0.156s
```
But with my INBOX  equivalent m4e query I am getting:

```
~/bin/mu$ time mu find maildir:/spinuvit/INBOX -t -r -s date --reverse -n 200 > tmp.txt

real    0m27.234s
user    0m26.300s
sys     0m0.946s
~/bin/mu$ time mu find maildir:/spinuvit/INBOX -t -r -s date -n 200 > tmp.txt

real    0m5.228s
user    0m4.756s
sys     0m0.472s
~/bin/mu$ time mu find maildir:/spinuvit/INBOX -t    -s date -n 200 > tmp.txt

real    0m5.269s
user    0m4.799s
sys     0m0.472s
~/bin/mu$ time mu find maildir:/spinuvit/INBOX    -r -s date -n 200 > tmp.txt

real    0m2.315s
user    0m1.981s
sys     0m0.332s
~/bin/mu$ time mu find maildir:/spinuvit/INBOX       -s date -n 200 > tmp.txt

real    0m0.070s
user    0m0.063s
sys     0m0.008s
~/bin/mu$ 
```

So `--reverse` adds a staggering 22 seconds!!

**Comment by lburton** [comment]
This is also hitting me. It makes mu / mu4e kind of unusable for me right now, unless I find some kind of workaround.

**Comment by lburton** [comment]
Oh, hold that thought. I just found that turning off mu4e-use-fancy-chars made a *huge* difference to performance. I need to find out what's up with that before blaming the index.

**Comment by jaksz** [comment]
I am also getting this problem. Around 40k messages total.

Inbox with <10 items needs a few seconds.

All mail needs a **lot** (20+ sec).

Maybe it's related to OS X? Those who are having this problem, what are you using?

**Comment by rpglover64** [comment]
I was on Linux (specifically, Debian) and I had this issue.

**Comment by tmalsburg** [comment]
@jaksz I'm on Xubuntu 14.04 (Debian derivative).

**Comment by jaksz** [comment]
@rpglover64 Did you get it resolved?

@tmalsburg Ok, that rules out the OS. Which leaves us the the mu4e implementation being at fault (as mu works just fine), right?

**Comment by rpglover64** [comment]
No. I just stopped trying to mirror my mail due to other reasons (gmail label woes).

**Comment by jaksz** [comment]
I should probably add that executing the the ```Last 7 days``` bookmark (which exists by default) is instantaneous as well.

However, creating a custom bookmark that would show ```maildir:inbox``` was as slow as jumping to the maildir.

Which narrows the issue down even more: querying by maildirs? Perhaps specifically with Gmail? @tmalsburg are you using Gmail?

**Comment by tmalsburg** [comment]
@jaksz I'm not anymore using Gmail because I switched to [Posteo](http://posteo.de/en), which I highly recommend.  (No more pseudo-free services for me if I can avoid it.)  But since mu operates on the local Maildir it's shouldn't really matter where the emails are coming from.

**Comment by jaksz** [comment]
Well, Gmail does folders in a non-standard way (views them as labels), maybe the problem is caused by this.

**Comment by lburton** [comment]
I switched to using the emacs nightlies (http://emacsformacosx.com) rather than railwaycat's build and although I lose things like momentum scrolling with the trackpad, it's *way* faster. I've left fancy chars disabled but been much happier with performance, including emacs as a whole.

**Comment by jaksz** [comment]
Are you saying that you can display your Inbox and it works instantaneously?

Luke Burton <notifications@github.com> (2015-08-18 01:25):

**Comment by lburton** [comment]
Yes, in this particular index I have about 100,000 messages, ~200 or so in the INBOX, and I don't get any appreciable delay when searching. This holds both for the rendering speed in emacs itself, as well as performing searches with mu on the command line.

**Comment by cpaulik** [comment]
I have the same problem with `fancy-chars` that @lburton reported. The thing is that it works snappy on my work machine with `fancy-chars` enabled but at home it is slowing it down a lot. So this must be an emacs issue or a OS problem. It works on Ubuntu14.04 but not on Arch.

**Comment by djcb** [comment]
Is some non-local filesystem involved (such as NFS, SMB)?

**Comment by cpaulik** [comment]
No my filesystem is local. 

I've also tried with an emacs 25 build that seemed to handle it way better so I guess that this is a problem with the emacs version I'm using. 

I just posted this in here since @lburton already started with the `fancy-chars` problem but I doubt that this is the same issue that OP reported.

**Comment by tmalsburg** [comment]
No non-local file-systems in my case either.  Everything is on my SSD.

I'm not sure whether the problem is in Emacs because when I run mu on the command line, I still get this pattern of processing times (`time mu find [options] > /dev/null)`:

|                  | related: no | related: yes |
|------------------|-------------|--------------|
| **threads: no**  | 1.6 s       | 3.8 s        |
| **threads: yes** | 7.9 s       | 10.2 s       |

So there seems to be substantial added cost for generating the threaded format.

Here are the times for opening my inbox (34k messages) in Mu4e (`(setq mu4e-headers-results-limit 500)`):

|                  | related: no | related: yes |
|------------------|-------------|--------------|
| **threads: no**  | < 1 s       | ~ 4 s        |
| **threads: yes** | ~ 5 s       | ~ 8 s        |

Tested on Ubuntu 14.04, recent development version of Emacs, current mu/mu4e from git (22b7022798adf3beb35fcb88611f8a567d29ac7b).  Please let me know if you need any other information or if you want me to run any other tests.

**Comment by djcb** [comment]
Hmmm,  seems @tmalsburg's numbers have something to do with mu, not mu4e.
```
% mu find --threads --include-related -n 500 maildir:/archive > /dev/null
mu find --threads --include-related -n 500  >   1,33s user 0,36s system 99% cpu 1,689 total
```
(there are ~20K messages in that folder)

When searching for related messages, we first run an unlimited query for all matching messages, then to a n=500 (or whatever) query for all the matching OR any with a message-id referred from one of the messages in the first query. Anyway, there's no reason why that should be so much slower in your case.

**Comment by djcb** [comment]
@tmalsburg: could you try commenting out lines 96-100 in mu-cmd-find.c, ie., like
```
	/* if (!mu_msg_is_readable (msg)) { */
	/* 	mu_msg_iter_next (iter); */
	/* 	return get_message (iter, after); */
	/* } */
```
and try your command-line tests again? That might give a hint.

**Comment by tmalsburg** [comment]
Just tried it and got very similar numbers:

|                  | related: no | related: yes |
|------------------|-------------|--------------|
| **threads: no**  | 0.04 s       | 4.5 s        |
| **threads: yes** | 6.1 s       | 10.4 s       |

**Comment by Adocentyn** [comment]
I'm on Arch Linux x64  and I had the same issue with mu/mu4e 0.9.11 and 0.9.13.

**Comment by stsquad** [comment]
On my corpus if ~800k messages it takes around 13 seconds to re-index although only only about 1.8 seconds to run find "find ~/Maildir/ -xtype f -and -readable" across the whole thing. The hardware is SSD based and I have enough RAM that everything should be cached. I can post a full perf dump if you want but is mainly looks like kernel RCU contention in the directory cache caused by:

                     - _MuStore::get_uid_term(char const*) const
and
                     - process_dir

**Comment by djcb** [comment]
@stsquad: that's interesting... I've optimized get_uid_term a bit (in master). Can you see if it makes things faster for you?

**Comment by stsquad** [comment]
That's certainly improved it a bit - down to about 9.6 seconds for the index. The get_uid_term has dropped off the main profile leaving "process_dir" as the main exerciser of sys_access/vfs_fstat which is again hitting the directory rcu:


[perf.hist.txt](https://github.com/djcb/mu/files/36479/perf.hist.txt)

**Comment by hinrik** [comment]
Responding to https://groups.google.com/forum/#!topic/mu-discuss/H-9_oo1pjFI, here are results from my 185k inbox, indexed with mu 37d5c12:

    find "": 4.0 s
    find -r "": 11.8 s
    find -t "": 45.7 s
    find -r - t "": 60.1 s

    find "flag:unread": 1.1 s
    find -r "flag:unread": 4.0 s
    find -t "flag:unread": 8.3 s
    find -r - t "flag:unread": 12.0 s

In mu4e, I have related messages enabled, fancy chars disabled, and full search disabled. Simple term searches are usually instant. Most of my bookmarks (typically 1 or more "list:" queries) open in under a second, but occasionally those can take up to 6 seconds, especially for mailing lists that have a lot of messages. My "flag:unread" bookmark almost always takes 12 seconds as described above. Since opening bookmarks is the most common action I do in mu4e, it would be nice to trim this query time down. Is this something mu can do, or is this an inherent limitation of a Xapian database of this size (185k msgs)?

The machine is:

Debian testing, on an ext4 filesystem (no NFS/SMB/etc)
Quad-core Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz
16GB RAM
2x 3TB 7200RPM SATA drives in RAID1

**Comment by andycowl** [comment]
Just started using mu/mu4e yesterday. mu and re-indexing is fine but I noted that an index rebuild of my 42,039 messages takes 106 seconds which seems slower than Dirk's sub-second re-indexing of 50k messages.

Hardware: Lenovo t420 (Intel dual core i5 CPU @ 2.50GHz, 8GB memory
O/S; RedHat Linux 7 Kernel 3.10.0-327.10.1.el7.x86_64
mu/mu4e : 0.9.17
Filesystem: :Local XFS

The other possible culprit is the McAffee disk encryption so I'll repeat the test on Linux Mint/ext4 filesystem with the same data set.

**Comment by tmalsburg** [comment]
Hi @andycowl, this sounds like an unrelated problem.  Perhaps you could open a new issue.

**Comment by djcb** [comment]
@andycowl: a full index rebuild is expected to take much longer than re-indexing, so no surprise there.

**Comment by andycowl** [comment]
Sorry - re-reading it, that comment was idiotic. Expecting 40k messages to be fully indexed in seconds. FWIW mu is twice as fast as 'notmuch' for my data set.

**Comment by nicolasavru** [comment]
I just sent out #1284, which makes the threading calculation scale linearly with maxnum. It doesn't improve performance when calculating threads for a large set of messages, but improve it significantly when calculating for a small set (say, the default maxnum=500).

Some time comparisons for a maildir with ~224k messages:
time mu find [options] "" > /dev/null

| master (4f6fd873fa82e899a03347321dc353e6ddbbac1a) | related: no | related: yes |
| ---------------------------------------------------------------------------- | -------------- | ---------------- |
| threads: no                                                                          | 4.50s         | 12.30s         |
| threads: no                                                                          | 46.64s       |   50.30s       |

| #1284  | related: no | related: yes |
| ---------------------------------------------------------------------------- | -------------- | ---------------- |
| threads: no |    4.47s               |   12.67s                |
| threads: no |    50.96s               |   58.70s                |


time mu find [options] -n 500 "" > /dev/null

| master (4f6fd873fa82e899a03347321dc353e6ddbbac1a) | related: no | related: yes |
| ---------------------------------------------------------------------------- | -------------- | ---------------- |
| threads: no                                                                          | 0.08s         |   28.80s       |
| threads: no                                                                          | 46.09s       |   64.56s       |

| #1284  | related: no | related: yes |
| ---------------------------------------------------------------------------- | -------------- | ---------------- |
| threads: no |    0.08s               |   0.09s                |
| threads: no |    0.12s               |   0.12s                |

As for speeding up the calculation without a maxnum... below is part of an execution time graph from kcachegrind (data from a valgrind callgrind run). The percentages are of the total run time. I don't have any great ideas on improving that at the moment. I considered putting the "thread paths" for a date descending ordering (likely the most common case) in the database, but adding new messages is a bit complicated (requires updating the database entries for the message's entire thread) and might be a performance hit during indexing operations. Or maybe it would be fine. I still don't love it though.

![2018-07-09-000942_1126x1387_scrot](https://user-images.githubusercontent.com/321994/42430354-b23ee58c-830c-11e8-8dce-939734be3f05.png)

**Comment by felixSchl** [comment]
Hey, what are the odds of getting #1284 merged? I cannot comment on the code, but I just tried it out and it's a total game changer in terms of usability for me: inboxes open pretty much instantly as opposed to 3-5 seconds later.

**Comment by markasoftware** [comment]
I am experiencing a similar issue on 1.0. My maildir has ~12,000 messages, and any `find` involving `--include-related` and/or `--threads` and no search term is pretty slow:

```
mark@way:~ $ time mu find --include-related -t -n 500 '' > /dev/null

real    0m1.332s
user    0m1.057s
sys     0m0.274s
```

That was tested on my laptop. It's about 3-5 seconds on my desktop. Interestingly, if I provide a search term it is nearly instant:

```
mark@way:~ $ time mu find --include-related -t -n 500 't' > /dev/null

real    0m0.077s
user    0m0.049s
sys     0m0.032s
```

Version:

```
mu (mail indexer/searcher) version 1.0
Copyright (C) 2008-2015 Dirk-Jan C. Binnema
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

I am using the openSUSE distribution provided binary. I downloaded the Ubuntu 18.10 binary and tested that too; it has identical performance.

**Comment by stsquad** [comment]
@markasoftware #1284 was merged into master - can you retest on that? It will be a while before the various distros update their packages.

**Comment by markasoftware** [comment]
@stsquad ah, I thought that it was in 1.0 because the PR was merged before February 3rd, but I guess it didn't make the cut. Building from source cuts down the search time to about 0.1s.

**Comment by nckx** [comment]
> It will be a while before the various distros update their packages.

Also, most distros package only upstream releases, not master.

**Comment by djcb** [comment]
Am going to assume this was fixed with the recent merge (thanks  @nicolasavru !) as mentioned. If there are remain issues after that, feel free to open new tickets, preferably with profile data (either emacs or e.g. 'perf' on linux)

**Comment by djcb** [comment]
(Note, #1238 was merged about a year after 1.0. Yes, we need a new release...)
